### PR TITLE
Methods for getting list of entities by their IDs

### PR DESCRIPTION
**Pull request by cuadradek**
- New methods "getEntitiesByIds" were created for Facility, Group, Member,
Resource, RichResource, User, UserExtSource and VO.

### CODE REVIEW

File: perun-core/src/main/java/cz/metacentrum/perun/core/api/FacilitiesManager.java [hunk 1]
```diff
@@ -59,6 +59,17 @@
 	 */
 	Facility getFacilityByName(PerunSession perunSession, String name) throws FacilityNotExistsException, PrivilegeException;
 
+	/**
+	 * Searches for the Facilities with specified ids.
+	 *
+	 * @param perunSession
+	 * @param ids
+	 * @return list of Facilities with specified ids
+	 * @throws InternalErrorException
```
Comment (metodej) [inline perun-core/src/main/java/cz/metacentrum/perun/core/api/FacilitiesManager.java:68]: I don't think there is any reason to include `InternalErrorException` to javadoc, could you please delete it from the javadocs you added?

**Review by metodej — APPROVED** [review]

File: perun-core/src/main/java/cz/metacentrum/perun/core/api/FacilitiesManager.java (hunk 1 above)
Comment (cuadradek) [reply perun-core/src/main/java/cz/metacentrum/perun/core/api/FacilitiesManager.java]: Done.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/FacilitiesManagerImpl.java [hunk 2]
```diff
@@ -336,6 +342,18 @@ public Facility getFacilityByName(PerunSession sess, String name) throws Facilit
 		}
 	}
 
+	@Override
+	public List<Facility> getFacilitiesByIds(PerunSession perunSession, List<Integer> ids) {
+		try {
+			SqlParameterSource parameters = new MapSqlParameterSource("ids", ids);
+
+			return namedParameterJdbcTemplate.query("select " + facilityMappingSelectQuery + " from facilities where id in (:ids)",
```
Comment (zlamalp) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/FacilitiesManagerImpl.java:350]: Using ``id in (...)`` is not effective for large number of IDs. Please see ``UsersManagerImpl.getUsersByIds()`` implementation. This apply to impl of all methods.

**Review by zlamalp — CHANGES_REQUESTED** [review]

File: perun-core/src/main/java/cz/metacentrum/perun/core/entry/UsersManagerEntry.java [hunk 3]
```diff
@@ -515,6 +515,20 @@ public UserExtSource getUserExtSourceById(PerunSession sess, int id) throws User
...
+		Utils.checkPerunSession(sess);
+
+		// Authorization
+		if (!AuthzResolver.authorizedInternal(sess, "getUserExtSourcesByIds_List<Integer>_policy")) {
+			throw new PrivilegeException(sess, "getUserExtSourcesByIds");
+		}
+		List<UserExtSource> userExtSources = getUsersManagerBl().getUserExtSourcesByIds(sess, ids);
+		userExtSources.removeIf(ues -> !AuthzResolver.authorizedInternal(sess, "filter-getUserExtSourcesByIds_List<Integer>_policy", ues));
+
+		return userExtSources;
```
Comment (HejdaJakub) [inline perun-core/src/main/java/cz/metacentrum/perun/core/entry/UsersManagerEntry.java:529]: Due to changes in perun-roles.yml you should modify this method.

**Review by HejdaJakub — CHANGES_REQUESTED** [review]

File: perun-base/src/main/resources/perun-roles.yml [hunk 4]
```diff
@@ -4529,6 +4661,28 @@ perun_policies:
...
+  filter-getUserExtSourcesByIds_List<Integer>_policy:
+    policy_roles:
+      - RPC:
+      - GROUPADMIN:
+      - GROUPOBSERVER:
+      - VOOBSERVER:
+      - VOADMIN:
+      - PERUNOBSERVER:
+    include_policies:
+      - default_policy
```
[the remark this answers is not in the document]
Comment (cuadradek) [reply perun-base/src/main/resources/perun-roles.yml]: I removed the filter policy.

File: perun-core/src/main/java/cz/metacentrum/perun/core/entry/UsersManagerEntry.java (hunk 3 above)
Comment (cuadradek) [reply perun-core/src/main/java/cz/metacentrum/perun/core/entry/UsersManagerEntry.java]: Done

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/FacilitiesManagerImpl.java (hunk 2 above)
Comment (cuadradek) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/FacilitiesManagerImpl.java]: Done

File: perun-core/src/main/java/cz/metacentrum/perun/core/implApi/GroupsManagerImplApi.java [hunk 5]
```diff
@@ -158,6 +158,16 @@
 	 */
 	Group getGroupByName(PerunSession perunSession, Vo vo, String name) throws GroupNotExistsException;
 
+	/**
+	 * Search for the groups with specified ids in all VOs.
```
Comment (zlamalp) [inline perun-core/src/main/java/cz/metacentrum/perun/core/implApi/GroupsManagerImplApi.java:162]: I would prefer having similar messages in javadoc for all those impl methods. Its ``get`` method. It doesn't search/find. Also you should note, that it silently skips non-existing entities.

File: perun-core/src/test/java/cz/metacentrum/perun/core/entry/UsersManagerEntryIntegrationTest.java [hunk 6]
```diff
@@ -1149,7 +1172,7 @@ public void getUsersByIds() throws Exception {
 			users.add(user2);
 		}
 
-		assertEquals(users, perun.getUsersManagerBl().getUsersByIds(sess, ids));
+		assertEquals(users, perun.getUsersManager().getUsersByIds(sess, ids));
```
Comment (zlamalp) [inline perun-core/src/test/java/cz/metacentrum/perun/core/entry/UsersManagerEntryIntegrationTest.java:1175]: When you are at it, can you please fix this test? When we use real DB, there is no guarantee, that returned list has same order as checked one, resulting in false positive test failure.
You can gather users on input to set and you can do the same for the result of getUsersByIds(). Then you can call assertEquals() on two sets, where order doesn't matter. Thank you.

File: perun-rpc/src/main/java/cz/metacentrum/perun/rpc/methods/FacilitiesManagerMethod.java [hunk 7]
```diff
@@ -42,6 +42,20 @@ public Facility call(ApiCaller ac, Deserializer parms) throws PerunException {
 		}
 	},
 
+	/*#
+	 * Searches for the Facilities with specified IDs.
```
Comment (zlamalp) [inline perun-rpc/src/main/java/cz/metacentrum/perun/rpc/methods/FacilitiesManagerMethod.java:46]: I would use unified javadoc message in RPC for all methods as I suggested for impl layer.

**Review by zlamalp — CHANGES_REQUESTED** [review]

**Review by HejdaJakub — APPROVED** [review]

File: perun-core/src/main/java/cz/metacentrum/perun/core/implApi/GroupsManagerImplApi.java (hunk 5 above)
Comment (cuadradek) [reply perun-core/src/main/java/cz/metacentrum/perun/core/implApi/GroupsManagerImplApi.java]: Done

File: perun-core/src/test/java/cz/metacentrum/perun/core/entry/UsersManagerEntryIntegrationTest.java (hunk 6 above)
Comment (cuadradek) [reply perun-core/src/test/java/cz/metacentrum/perun/core/entry/UsersManagerEntryIntegrationTest.java]: Done

File: perun-rpc/src/main/java/cz/metacentrum/perun/rpc/methods/FacilitiesManagerMethod.java (hunk 7 above)
Comment (cuadradek) [reply perun-rpc/src/main/java/cz/metacentrum/perun/rpc/methods/FacilitiesManagerMethod.java]: Done

File: perun-core/src/test/java/cz/metacentrum/perun/core/entry/UsersManagerEntryIntegrationTest.java (hunk 6 above)
Comment (zlamalp) [reply perun-core/src/test/java/cz/metacentrum/perun/core/entry/UsersManagerEntryIntegrationTest.java]: Thank you.

**Review by zlamalp — APPROVED** [review]

**Pull request merged**